spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.java

```diff
@@ -633,9 +633,15 @@ private void considerAddingAConsumer() {
 		}
 	}
 
+	private int getConsumersThatHaveNotInitiatedStopping() {
+		synchronized (this.consumersMonitor) {
+			return (int) this.consumers.values().stream().filter(x -> x = Boolean.TRUE).count();
```

Comment:
Hello, @zaharidichev !

Thank you for contribution first of all!

Well, let me review your PR and share my vision.

Your `filter()` looks wrong. It always assigns `Boolean.TRUE` to the `x` and returns that `true` 😄 .
OTOH the issue is a bug and we must back port it to versions which aren't support Java 8 yet.

One more thought that it looks like we don't need this method at all:
- leave `this.consumers.size() > this.concurrentConsumers` condition as is;
- instead of `this.consumers.put(consumer, false)` use `this.consumers.remove(consumer)`;
- the `isActive(BlockingQueueConsumer)` logic is based on the fact that a `value` for the `consumer` can be `null`, so we are safe to just remove `consumer` from the cache and don't care about its state;
- the `AsyncMessageProcessingConsumer.run()` should be free from of the `synchronized (SimpleMessageListenerContainer.this.consumersMonitor) {` block as well, since we don't need it here any more.

That way we will be able to avoid such kind of hell iteration in the `Map` on each `considerStoppingAConsumer()` call.
And at the same time we still reach atomicity for the stopping consumers.

Looks like we have to change `setConcurrentConsumers()`  to use `this.consumers.remove(consumer)` as well, because `SimpleMessageListenerContainerLongTests` fails for me locally with my logic. 

>  I mean surely they can still be processing messages.

Yes, indeed. That is a main point why even being shutting down we still treat it as an active because any "in-flight" message might be in process yet. Even direct `stop()` doesn't prevent that behavior. But at least it is some middleground between nothing and real state. Anyway externally you can't be sure how an internal state may be changed in between your actions. That's why we should ensure the proper state management internally to prevent any unexpected involvements. 

So, let's fix the current problem and leave `activeCount` as is or feel free to start a new JIRA for some argument to reconsider it.

At the same time we still can ask @garyrussell for his thoughts on the matter.

Thanks